pkg/reconciler/pipeline/dag/dag.go

```diff
@@ -121,40 +126,62 @@ func GetCandidateTasks(g *Graph, doneTasks ...string) (sets.String, error) {
 }
 
 func linkPipelineTasks(prev *Node, next *Node) error {
-	// Check for self cycle
-	if prev.Task.HashKey() == next.Task.HashKey() {
-		return fmt.Errorf("cycle detected; task %q depends on itself", next.Task.HashKey())
-	}
-	// Check if we are adding cycles.
-	path := []string{next.Task.HashKey(), prev.Task.HashKey()}
-	if err := lookForNode(prev.Prev, path, next.Task.HashKey()); err != nil {
-		return fmt.Errorf("cycle detected: %w", err)
-	}
 	next.Prev = append(next.Prev, prev)
 	prev.Next = append(prev.Next, next)
 	return nil
 }
 
-func lookForNode(nodes []*Node, path []string, next string) error {
-	for _, n := range nodes {
-		path = append(path, n.Task.HashKey())
-		if n.Task.HashKey() == next {
-			return errors.New(getVisitedPath(path))
+// use Kahn's algorithm to find cycles in dependencies
+func findCyclesInDependencies(deps map[string][]string) error {
```

Comment:
This function wrongly errors on:
```go
deps := map[string][]string{
  "a": []string{},
  "b": []string{"c", "d"},
  "c": []string{},
  "d": []string{},
}
// valid graph, topo-order: a, c, d, b
// but the call erros: `task "c" depends on `
```

Reply:
@Udiknedormin thanks for pointing. I have guarded the code to be safe in such case, even though graph builder does not include tasks with empty dependencies.